sound/soc/sof/ipc4-topology.h

```diff
@@ -148,13 +148,15 @@ struct ipc4_pipeline_set_state_data {
  * struct sof_ipc4_available_audio_format - Available audio formats
  * @base_config: Available base config
  * @out_audio_fmt: Available output audio format
+ * @input_audio_fmts: Available input audio formats
  * @ref_audio_fmt: Reference audio format to match runtime audio format
  * @dma_buffer_size: Available Gateway DMA buffer size (in bytes)
  * @audio_fmt_num: Number of available audio formats
  */
 struct sof_ipc4_available_audio_format {
 	struct sof_ipc4_base_module_cfg *base_config;
 	struct sof_ipc4_audio_format *out_audio_fmt;
+	struct sof_ipc4_audio_format *input_audio_fmts;
```

Comment:
@plbossart storing input formats as part of the base_cfg is not intuitive. Thats exactly why you think it is rocket science. I've moved it out of the base_config and called i input_audio_fmts. Can I please request you to be a little patient and walk through the whole patch set? In the end you will end up with the clean struct for available_audio_format

Reply:
put yourself in my shoes @ranj063. I just spent *HOURS* to figure out how this works, and the BASIC definition we have this is:

````c
struct sof_ipc4_base_module_cfg {
	uint32_t cpc; /* the max count of Cycles Per Chunk processing */
	uint32_t ibs; /* input Buffer Size (in bytes)  */
	uint32_t obs; /* output Buffer Size (in bytes) */
	uint32_t is_pages; /* number of physical pages used */
	struct sof_ipc4_audio_format audio_fmt;
} __packed __aligned(4);
````

that's also part of the documentation.

So no, I am sorry, I can't deal with new structures that revisit the conventions used so far. This is NOT aligned with documentation and not aligned with firmware definitions.

Reply:
@plbossart what part is not aligned with documentation? The base_module_cfg is still following the same definition as documentation. There is no hint in the documentation about base_module_cfg being an array in the first place. The way we're parsing the audio formats today just wont work as it is only meant for pin 0. What do we do with the other pin formats then? I think have all available formats with all the pin formats parsed from topology in shot makes a lot more sense than what we have today. 

And Im sorry you spent hours understanding this and that is precisely why I redid this part so that the next person looking at this code wont have to jump through the same hoops.